edb/server/protocol/auth_ext/base.py

```diff
@@ -37,14 +43,98 @@ def __init__(
         self.client_secret = client_secret
         self.http_factory = http_factory
 
-    def get_code_url(self, state: str, redirect_uri: str) -> str:
+    async def get_code_url(self, state: str, redirect_uri: str) -> str:
         raise NotImplementedError
 
-    async def exchange_code(self, code: str) -> str:
+    async def exchange_code(self, code: str) -> data.OAuthAccessTokenResponse:
         raise NotImplementedError
 
-    async def fetch_user_info(self, token: str) -> data.UserInfo:
+    async def fetch_user_info(
+        self, token_response: data.OAuthAccessTokenResponse
+    ) -> data.UserInfo:
         raise NotImplementedError
 
     def _maybe_isoformat_to_timestamp(self, value: str | None) -> float | None:
         return datetime.fromisoformat(value).timestamp() if value else None
+
+
+class OpenIDProvider(BaseProvider):
+    def __init__(self, name: str, issuer_url: str, *args, **kwargs):
+        super().__init__(name, issuer_url, *args, **kwargs)
+
+    async def get_code_url(self, state: str, redirect_uri: str) -> str:
+        oidc_config = await self._get_oidc_config()
+        params = {
+            "client_id": self.client_id,
+            "scope": "openid profile email",
+            "state": state,
+            "redirect_uri": redirect_uri,
+            "nonce": str(uuid.uuid4()),
+            "response_type": "code",
+        }
+        encoded = urllib.parse.urlencode(params)
+        return f"{oidc_config.authorization_endpoint}?{encoded}"
+
+    async def exchange_code(
+        self, code: str
+    ) -> data.OpenIDConnectAccessTokenResponse:
+        oidc_config = await self._get_oidc_config()
+
+        token_endpoint = urllib.parse.urlparse(oidc_config.token_endpoint)
+        async with self.http_factory(
+            base_url=f"{token_endpoint.scheme}://{token_endpoint.netloc}"
+        ) as client:
+            resp = await client.post(
+                token_endpoint.path,
+                json={
+                    "grant_type": "authorization_code",
+                    "code": code,
+                    "client_id": self.client_id,
+                    "client_secret": self.client_secret,
+                },
+            )
+            json = resp.json()
+
+            return data.OpenIDConnectAccessTokenResponse(**json)
+
+    async def fetch_user_info(
+        self, token_response: data.OAuthAccessTokenResponse
+    ) -> data.UserInfo:
+        if not isinstance(
+            token_response, data.OpenIDConnectAccessTokenResponse
+        ):
+            raise TypeError(
+                "token_response must be of type "
+                "OpenIDConnectAccessTokenResponse"
+            )
```

Comment:
I battled with ChatGPT for a while trying to figure out a way to restrict this at compile time such that callers would need to pass `OpenIDConnectAccessTokenResponse` here (which in practice we do since we just do `exchange_code -> fetch_user_info`), but this is all I could come up with. Open to suggestions on a better way to do this!

Reply:
> If we are sure/confident of the given type of response, I'd say using assert is also cool here as it is semantically for debugging (asserts are not run under python -O, though I don't think we are using that option).

The thing I was hoping to find was some way for the type system to catch if we accidentally broke the type of `exchange_code` in the OpenID Connect providers (e.g. by accidentally using `data.OAuthAccesTokenResponse(**json)`). Or if we did some other refactor that accidentally broke our contract here.

It's fine if the answer is: there is nothing the type system can do here to prevent the issue, so all solutions will be some runtime check. If so, I'll just leave what I have.